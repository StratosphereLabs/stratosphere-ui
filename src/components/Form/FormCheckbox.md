# FormCheckbox

The `FormCheckbox` component is a custom form field component built with React. It provides a styled checkbox input that integrates with the DaisyUI library, and includes support for the `react-hook-form` library.

## Props

The `FormCheckbox` component accepts the following props:

- `children?: React.ReactNode`: Optional. Any additional content to be displayed alongside the checkbox.
- `className?: string`: Optional. Additional class names to be applied to the component.
- `color?: string`: Optional. The color of the checkbox. Defaults to 'ghost'.
- `controllerProps?: Omit<UseControllerProps<Values>, 'name'>`: Optional. Props to pass to the `useController` hook, which manages the form input's state.
- `inputRef?: RefObject<HTMLInputElement>`: Optional. A ref to the input element of the checkbox.
- `isRequired?: boolean`: Optional. Indicates whether the field is required.
- `labelText?: string`: Optional. The label text for the checkbox.
- `name: Path<Values>`: Required. The name of the field within the form.
- `showDirty?: boolean`: Optional. Indicates whether the field should be styled to indicate if it has been modified.

## Usage

To use the `FormCheckbox` component, simply import it into your React component and render it with the appropriate props:

```tsx
import { useForm } from 'react-hook-form';
import { Form, FormCheckbox } from 'stratosphere-ui';

interface FormValues {
  exampleCheckbox: boolean;
}

export const MyForm = () => {
  const methods = useForm<FormValues>();
  const onSubmit = (formData: FormValues) => {
    console.log(formData);
  };
  return (
    <Form methods={methods} onFormSubmit={onSubmit}>
      <FormCheckbox name="exampleCheckbox" labelText="Example Checkbox" />
    </Form>
  );
};
```
