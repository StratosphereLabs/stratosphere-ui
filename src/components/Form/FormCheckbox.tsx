import classNames from 'classnames';
import { RefObject } from 'react';
import { Checkbox, CheckboxProps } from 'react-daisyui';
import { FieldValues, useController } from 'react-hook-form';
import { FormLabel } from './FormLabel';
import { FormFieldProps } from './types';
import { useFieldColor } from '../../hooks';

export interface FormCheckboxProps<Values extends FieldValues>
  extends Omit<FormFieldProps<Values>, 'placeholder'>,
    Omit<CheckboxProps, 'name'> {
  inputRef?: RefObject<HTMLInputElement>;
}

export const FormCheckbox = <Values extends FieldValues>({
  children,
  className,
  color,
  controllerProps,
  inputRef,
  isRequired,
  labelText,
  name,
  showDirty,
  ...props
}: FormCheckboxProps<Values>): JSX.Element => {
  const { field } = useController({ name, ...controllerProps });
  const fieldColor = useFieldColor(name, showDirty);
  return (
    <div className={classNames('flex items-center gap-2', className)}>
      <Checkbox
        {...field}
        {...props}
        color={fieldColor ?? color ?? 'ghost'}
        ref={inputRef}
      />
      <div className="flex flex-col">
        {labelText !== undefined ? (
          <FormLabel htmlFor={props.id} isRequired={isRequired}>
            {labelText}
          </FormLabel>
        ) : null}
        {children}
      </div>
    </div>
  );
};
