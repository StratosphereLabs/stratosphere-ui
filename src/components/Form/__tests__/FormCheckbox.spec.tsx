import { render, screen } from '@testing-library/react';
import { FormCheckbox } from '../FormCheckbox';
import FormProvider from './FormProvider';

describe('FormCheckbox', () => {
  it('renders component', () => {
    const { container } = render(
      <FormProvider defaultValues={{ checkbox: false }}>
        <FormCheckbox name="checkbox" />
      </FormProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders the label and children', () => {
    const labelText = 'Agree to terms';
    const childrenText = 'I agree to the terms and conditions';
    render(
      <FormProvider defaultValues={{ checkbox: false }}>
        <FormCheckbox labelText={labelText} name="checkbox">
          <span>{childrenText}</span>
        </FormCheckbox>
      </FormProvider>,
    );
    expect(screen.getByText(labelText)).toBeInTheDocument();
    expect(screen.getByText(childrenText)).toBeInTheDocument();
  });

  it('renders an asterisk if isRequired is true', () => {
    const labelText = 'Agree to terms';
    render(
      <FormProvider defaultValues={{ checkbox: false }}>
        <FormCheckbox labelText={labelText} isRequired name="checkbox" />
      </FormProvider>,
    );
    expect(screen.getByText('*')).toBeInTheDocument();
  });
});
