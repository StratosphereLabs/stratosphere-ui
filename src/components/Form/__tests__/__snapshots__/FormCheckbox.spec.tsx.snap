// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormCheckbox > renders component 1`] = `
<div>
  <div
    class="flex items-center gap-2"
  >
    <input
      class="checkbox checkbox-ghost"
      name="checkbox"
      type="checkbox"
      value="false"
    />
    <div
      class="flex flex-col"
    />
  </div>
</div>
`;
